### PR TITLE
✨ Support mapping local and remote Capsule traffic using tunneling

### PR DESCRIPTION
**Pull request by andersjohnsen**

### CODE REVIEW

**Review by Munklinde96 — APPROVED** [review]

File: pkg/api/platform/v1/dev_types.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+	TCP bool `json:"tcp,omitempty" protobuf:"1"`
+	// AllowOrigin sets the `Access-Control-Allow-Origin` Header on responses to
+	// the provided value, allowing local by-pass of CORS rules.
+	// Ignored if TCP is enabled.
+	AllowOrigin string `json:"allowOrigin,omitempty" protobuf:"2"`
+	// ChangeOrigin changes the Host header to match the given target. If not set,
+	// the Host header will be that of the original request.
+	// This does not impact the Origin header - use `Headers` to set that.
+	// Ignored if TCP is enabled.
+	ChangeOrigin bool `json:"changeOrigin,omitempty" protobuf:"3"`
```
Comment (MatiasFrank) [inline pkg/api/platform/v1/dev_types.go:70]: Do we copy over the headers of the incoming HTTP request? That is, if the incoming HTTP request has it's Origin set, will that origin persist into the cluster?

File: pkg/api/platform/v1/dev_types.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+type HostNetwork struct {
+	// HostInterfaces are interfaces activated on the local machine (the host) and forwarded
+	// to the Kubernetes cluster capsules.
+	HostInterfaces []HostInterface `json:"hostInterfaces" protobuf:"1"`
+
+	// CapsuleInterfaces are interfaces activated on the Capsule within the Kubernetes cluster
+	// and forwarded to the local machine (the host). The traffic is directed to a single target,
+	// e.g. `localhost:8080`.
+	CapsuleInterface []CapsuleInterface `json:"capsuleInterfaces" protobuf:"2"`
```
Comment (MatiasFrank) [inline pkg/api/platform/v1/dev_types.go:37]: Why are they two different fields? Couldn't we have a single `Interface` config which maps a specific (host port, capsule interface) pair and thus encapsulates both `HostInterface` and `CapsuleInterface`?

File: cmd/rig/cmd/dev/host.go [hunk 3]
```diff
@@ -0,0 +1,418 @@
...
+
+	tunnelID atomic.Uint64
+	lock     sync.Mutex
+	tunnels  map[uint64]*tunnel.Buffer
+}
+
+func (t *ReverseTunnel) Run(ctx context.Context) error {
+	for port, listener := range t.hostInterfaces {
+		port := port
+		listener := listener
```
Comment (MatiasFrank) [inline cmd/rig/cmd/dev/host.go:292]: This have been fixed in go 1.22 (finally)

File: cmd/rig/cmd/dev/host.go [hunk 4]
```diff
@@ -0,0 +1,418 @@
...
+	return capsule_cmd.PortForwardOnListener(
+		ctx, c.Rig, c.Scope, capsuleName, instanceID, l, cfg.GetNetwork().GetTunnelPort(), true)
+}
+
+type hostListener struct {
+	cfg      *platformv1.HostInterface
+	listener net.Listener
+}
+
+type ReverseTunnel struct {
```
Comment (MatiasFrank) [inline cmd/rig/cmd/dev/host.go:279]: Naming wise, isn't it just a tunnel? Considering it handles both host and capsule interfaces it doesn't have a preferred direction?

File: pkg/tunnel/tunnel.go [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+				return nil
+			},
+		}
+
+		c1, c2 := net.Pipe()
+		l := newSingleConnListener(c2)
+		go func() {
+			_ = http.Serve(l, rp)
+		}()
+		conn = c1
```
Comment (MatiasFrank) [inline pkg/tunnel/tunnel.go:148]: What was the purpose of the piping?

File: pkg/api/platform/v1/dev_types.go [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+	// TunnelPort for which the proxy-capsule should listen on. This is automatically set by the tooling.
+	TunnelPort uint32 `json:"tunnelPort,omitempty" protobuf:"3"`
+}
+
+type HostInterface struct {
+	// Port on the host from where to accept traffic from.
+	Port *uint32 `json:"port,omitempty" protobuf:"2"`
+	// CapsuleTarget is the capsule-name:capsule-port to forward traffic to.
+	CapsuleTarget string           `json:"capsuleTarget" protobuf:"1"`
```
Comment (MatiasFrank) [inline pkg/api/platform/v1/dev_types.go:47]: Why does `CapsuleTarget` contain a capsule name? Isn't it always the same capsule we're trying to hit within one `HostCapsule`?

File: pkg/api/platform/v1/dev_types.go (hunk 1 above)
Comment (andersjohnsen) [reply pkg/api/platform/v1/dev_types.go:70]: Yes, unless Header is overwritten

File: cmd/rig/cmd/dev/host.go [hunk 7]
```diff
@@ -0,0 +1,418 @@
...
+
+		_, portStr, err := net.SplitHostPort(l.Addr().String())
+		if err != nil {
+			return err
+		}
+
+		port, err := strconv.ParseUint(portStr, 10, 32)
+		if err != nil {
+			return err
+		}
```
[the remark this answers is not in the document]
Comment (andersjohnsen) [reply cmd/rig/cmd/dev/host.go:236]: I can be dynamic (e.g 0 -> 31284)

File: cmd/rig/cmd/dev/host.go (hunk 4 above)
Comment (andersjohnsen) [reply cmd/rig/cmd/dev/host.go:279]: It is :)

File: cmd/rig/cmd/dev/host.go [hunk 8]
```diff
@@ -0,0 +1,418 @@
...
+		res, err := t.tunnelStream.Recv()
+		if err != nil {
+			return err
+		}
+
+		switch v := res.GetMessage().GetMessage().(type) {
+		case *api_tunnel.TunnelMessage_NewTunnel:
+			tunnelID := v.NewTunnel.GetTunnelId()
+			buff := tunnel.NewBuffer()
+
```
[the remark this answers is not in the document]
Comment (andersjohnsen) [reply cmd/rig/cmd/dev/host.go:320]: It's done through the shared library, in NewTunnel

File: pkg/api/platform/v1/dev_types.go (hunk 6 above)
Comment (andersjohnsen) [reply pkg/api/platform/v1/dev_types.go:47]: As discussed offline :)

**Pull request merged**